extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts

```diff
@@ -0,0 +1,204 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { JupyterServerInstallation } from '../../jupyter/jupyterServerInstallation';
+import { IPackageManageProvider, IPackageDetails } from '../../types';
+
+export interface ManagePackageDialogOptions {
+	multiLocations: boolean;
+	defaultLocation?: string;
+}
+
+/**
+ * Manage package dialog model
+ */
+export class ManagePackageDialogModel {
+
+	public currentProvider: string;
+	private _locations: Set<string> = new Set<string>();
+	private _packageTypes: Map<string, IPackageManageProvider[]> = new Map<string, IPackageManageProvider[]>();
+
+	/**
+	 * Creates new instance of the model
+	 * @param _jupyterInstallation Jupyter installation
+	 * @param _packageManageProviders package manage providers
+	 * @param _options dialog options
+	 */
+	constructor(
+		private _jupyterInstallation: JupyterServerInstallation,
+		private _packageManageProviders: Map<string, IPackageManageProvider>,
+		private _options?: ManagePackageDialogOptions) {
+
+		if (!this._packageManageProviders || this._packageManageProviders.size === 0) {
+			throw Error('Invalid list of package manager providers');
+		}
+	}
+
+	/**
+	 * Initialized the model
+	 */
+	public async init(): Promise<void> {
+		await this.loadCaches();
+		this.loadOptions();
+	}
+
+	/**
+	 * Loads the model options
+	 */
+	private loadOptions(): void {
+		// Set Default Options
+		//
+		if (!this._options) {
+			this._options = {
+				multiLocations: true,
+				defaultLocation: undefined
+			};
+		}
+
+		if (this._options.defaultLocation && !this._packageTypes.get(this._options.defaultLocation)) {
+			throw new Error(`Invalid default location '${this._options.defaultLocation}`);
+		}
+
+		if (!this._options.multiLocations && !this.defaultLocation) {
+			throw new Error('Default location not specified for single location mode');
+		}
+	}
+
+	/**
+	 * Returns the providers map
+	 */
+	public get packageManageProviders(): Map<string, IPackageManageProvider> {
+		return this._packageManageProviders;
+	}
+
+	/**
+	 * Returns the current provider
+	 */
+	public get currentPackageManageProvider(): IPackageManageProvider {
```

Comment:
since it can return undefined, the return type should be IPackageManageProvider | undefined

Reply:
why? any option can be undefined too right? 

---
In reply to: [350545075](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350545075) [](ancestors = 350545075)

Reply:
It depends. If strict null checking is enabled then typescript will enforce this kind of behavior - so if you say you return IPackageManageProvider then it could ONLY return something implementing that. So this would be an error since undefined isn't an instanceof IPackageManageProvider.

Now that being said we don't currently have strict null checks enabled for the entire code base. It's been something we (mostly Anthony) have been working on but it's a lot of work fixing up our code to do that.

But we should still try to adhere by those rules whenever we can so it's very clear when something can possibly be undefined since that can save us having to do unnecessary undefined checks in some scenarios.

Reply:
got it. thanks for clarifying that. 

---
In reply to: [354075140](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r354075140) [](ancestors = 354075140)